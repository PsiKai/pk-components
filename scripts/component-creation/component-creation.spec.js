import { execSync } from "child_process"
import fs from "fs"
import path from "path"
import {
  generateIndexFile,
  generateComponentFile,
  generateModelFile,
  generateSpecFile,
  generateReadmeFile,
} from "./component-creation-templates"

describe("component-creation", () => {
  const componentName = "TestComponent"
  const filePath = path.join(__dirname, `../../src/components/${componentName}`)

  beforeEach(() => {
    execSync(`npm run component ${componentName}`)
  })

  afterEach(() => {
    if (fs.existsSync(filePath)) {
      fs.rmdirSync(filePath, { recursive: true })
    }
  })

  describe("with file system", async () => {
    it("should create the directory in the components folder", async () => {
      expect(fs.existsSync(filePath)).toBe(true)
    })
    it("should create the index.ts file", async () => {
      expect(fs.existsSync(`${filePath}/index.ts`)).toBe(true)
    })
    it("should create the component file", async () => {
      expect(fs.existsSync(`${filePath}/${componentName}.tsx`)).toBe(true)
    })
    it("should create the model file", async () => {
      expect(fs.existsSync(`${filePath}/${componentName}.model.ts`)).toBe(true)
    })
    it("should create the css file", async () => {
      expect(fs.existsSync(`${filePath}/${componentName}.css`)).toBe(true)
    })
    it("should create the component spec file", async () => {
      expect(fs.existsSync(`${filePath}/${componentName}.spec.tsx`)).toBe(true)
    })
    it("should create the readme file", async () => {
      expect(fs.existsSync(`${filePath}/README.md`)).toBe(true)
    })
  })

  describe("with existing component of same name", () => {
    it("should throw an error", async () => {
      expect(() => execSync(`npm run component ${componentName}`)).toThrow(
        `ERROR\nCOMPONENT ALREADY EXISTS: ${componentName}`,
      )
    })
  })

  describe("with no component name", () => {
    it("should throw an error", async () => {
      expect(() => execSync("npm run component")).toThrow("ERROR: Please provide a component name")
    })
  })
})

describe("component-creation template methods", () => {
  const componentName = "TestComponent"

  describe("generateComponentFile", () => {
    it("should return string content for index file", () => {
      expect(generateIndexFile(componentName)).toMatchSnapshot()
    })
  })

  describe("generateComponentFile", () => {
    it("should return string content for component file", () => {
      expect(generateComponentFile(componentName)).toMatchSnapshot()
    })
  })

  describe("generateModelFile", () => {
    it("should return string content for model file", () => {
      expect(generateModelFile(componentName)).toMatchSnapshot()
    })
  })

  describe("generateSpecFile", () => {
    it("should return string content for spec file", () => {
      expect(generateSpecFile(componentName)).toMatchSnapshot()
    })
  })

  describe("generateReadmeFile", () => {
    it("should return string content for readme file", () => {
      expect(generateReadmeFile(componentName)).toMatchSnapshot()
    })
  })
})
