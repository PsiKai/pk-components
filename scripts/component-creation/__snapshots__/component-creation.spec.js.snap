// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component-creation template methods > generateComponentFile > should return string content for component file 1`] = `
"import { TTestComponentProps } from "./TestComponent.model"
import "./TestComponent.css"

export function TestComponent(props: TTestComponentProps) {
  return <div>TestComponent</div>
}
"
`;

exports[`component-creation template methods > generateComponentFile > should return string content for index file 1`] = `
"export { TestComponent } from "./TestComponent"
export type { TTestComponentProps } from "./TestComponent.model"
"
`;

exports[`component-creation template methods > generateModelFile > should return string content for model file 1`] = `
"import { AllHtmlAttributes } from "../../core-types"

export type TTestComponentProps = AllHtmlAttributes & {
  className?: string
  children?: React.ReactNode
}
"
`;

exports[`component-creation template methods > generateReadmeFile > should return string content for readme file 1`] = `
"### TestComponent Component

#### Description

A brief description of the component. Basic functionality and behavior.

#### Props

| Prop Name            | Type                                                                | Required | Default       | Description                                     |
| -------------------- | ------------------------------------------------------------------- | -------- | ------------- | ----------------------------------------------- |
| \`[htmlAttributes]\` | \`React.AllHTMLAttributes<HTMLElement>\`                            | No       | \`undefined\` | Any valid HTML attribute for the element type   |
| \`aria-*\`          | \`[key: aria-\${string}]: string \\| number \\| boolean \\| null\` | No       | \`undefined\` | Optional Accessibility attributes               |
| \`data-*\`          | \`[key: data-\${string}]: string \\| number \\| boolean \\| null\` | No       | \`undefined\` | Optional dataset attributes                     |
| \`className\`        | \`string\`                                                          | No       | \`undefined\` | Additional class names to apply to the spinner. |

#### Example

\`\`\`tsx
import { TestComponent } from "pk-components"

function YourComponent() {
  return <TestComponent>Your content here</TestComponent>
}
\`\`\`

[Live Demo](https://psikai.github.io/pk-components#TestComponent)
"
`;

exports[`component-creation template methods > generateSpecFile > should return string content for spec file 1`] = `
"import React from "react"
  import { render, screen } from "@testing-library/react"
  import "@testing-library/jest-dom"

  import { TestComponent } from "./TestComponent"

  describe("TestComponent", () => {
    beforeEach(() => {
      render(<TestComponent />)
    })

    it("should render the component", () => {
      const component = screen.getByText("TestComponent")
      expect(component).toBeInTheDocument()
    })
  })
"
`;
